docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+# The Module Lifecycle
+
+This design note is intended for people who are somewhat familiar with Modules but want to understand more about the design principles behind the abstraction. After reading this, you should have a solid understanding of the assumptions and guarantees the Module API is built upon. If you have no practical experience with Modules yet, Please take a look at the [MNIST tutorial](https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html).
+
+Modules are the core of Flax. This is a practical guide to the design and behavior of the `Module` class. By the end, you should feel comfortable to go off the beaten track and use Modules in new ways.
+
+
+## Overview
+
+### Definition
+
+Let’s start with a high-level overview of the Module lifecycle. We start by defining a simple Module:
+
+
+```python
+class MLP(nn.Module):
+  # 1. Attribute annotations
+  hidden_size: int
+  out_size: int
+
+  # 2. The `setup` method
+  def setup(self):
+    self.hidden = nn.Dense(self.hidden_size)
+    self.out = nn.Dense(self.out_size)
+
+  # 3. User methods  
+  def __call__(self, x):
+    a = self.hidden(x)
+    h = nn.relu(a)
+    return self.out(h)
+```
+
+
+This Module consists of:
+
+
+1. **_Attribute annotations_**, defined as [dataclass](https://docs.python.org/3/library/dataclasses.html) fields. These annotations automatically define a constructor.
+2. **The `setup` method,** that creates submodules and assigns them to attributes.
+3. A **user-defined method**. By convention, most Modules have just one  `__call__` method, but you can define multiple methods or use different method names.
+
+### Construction/initialization
+
+Now we want to construct and use the `MLP` Module:
+
+
+```python
+mlp = MLP(5, 3)
```

Comment:
In terms of style, should we encourage passing the arguments as keywords here?

Reply:
done